File: ios/Classes/VideoRawDataController.m
#import <Foundation/Foundation.h>
#import "VideoRawDataController.h"
#import <AgoraRtcKit/AgoraRtcKit.h>
#import <CoreVideo/CoreVideo.h>
#import <CoreMedia/CoreMedia.h>
#import <objc/message.h>

// ARCHITECTURE NOTE
//
// This file used to run the INVERTED model: Agora owned the capture device and
// handed each captured frame to Nosmai via -onCaptureVideoFrame: /
// +processExternalPixelBuffer:shouldFlip:. That was structurally broken for two
// reasons, so it has been removed rather than finished:
//   1. It forced the SDK into NosmaiProcessingModeOffscreen, whose setter calls
//      stopProcessing internally and tears down the AVCaptureSession — preview
//      and streaming could therefore never be live at the same time.
//   2. It blocked Agora's capture thread on a dispatch_semaphore for up to 50ms
//      per frame while waiting for the filtered result.
//
// The model below matches the shipped Android bridge: Nosmai owns the camera and
// PUSHES filtered frames into Agora as an external video source.

// Encoder parameters — MUST stay identical to the Android bridge
// (VideoRawDataController.kt:24-27, :293-306). Divergence here shows up as a
// different aspect/quality between the two platforms for the same broadcast.
static const NSInteger kStreamWidth = 720;
static const NSInteger kStreamHeight = 1280;
static const NSInteger kStreamBitrateKbps = 2500;
static const NSInteger kStreamMinBitrateKbps = 1200;

// AgoraVideoFrame.format value for a CVPixelBufferRef carried in `textureBuf`
// (AgoraObjects.h:1792 — "12: iOS texture (CVPixelBufferRef)").
static const NSInteger kAgoraVideoFormatCVPixelBuffer = 12;

#pragma mark - Nosmai plugin reflection

// This bridge does NOT link the Nosmai framework (nosmai_agora_bridge.podspec
// depends only on Flutter + AgoraRtcEngine_iOS). The camera plugin lives in a
// separate pod, so every call into it goes through runtime reflection — the same
// contract the previous implementation used.
static Class NosmaiPluginClass(void) {
    return NSClassFromString(@"NosmaiFlutterPlugin");
}

// Arms Nosmai's live-frame output: Nosmai keeps ownership of the camera and
// pushes every filtered frame to us. Declared in
// camsdk_v2/ios/Classes/NosmaiFlutterPlugin.h (+setLiveStreamFrameCallback:).
static BOOL ArmNosmaiLiveFrames(void (^callback)(CVPixelBufferRef, double)) {
    Class cls = NosmaiPluginClass();
    SEL sel = @selector(setLiveStreamFrameCallback:);
    if (!cls || ![cls respondsToSelector:sel]) {
        return NO;
    }
    void (*fn)(Class, SEL, id) = (void (*)(Class, SEL, id))objc_msgSend;
    fn(cls, sel, callback);
    return YES;
}

static void DisarmNosmaiLiveFrames(void) {
    Class cls = NosmaiPluginClass();
    SEL sel = @selector(clearLiveStreamFrameCallback);
    if (cls && [cls respondsToSelector:sel]) {
        void (*fn)(Class, SEL) = (void (*)(Class, SEL))objc_msgSend;
        fn(cls, sel);
    }
}

@interface VideoRawDataController ()<AgoraRtcEngineDelegate>

@property(nonatomic, strong) AgoraRtcEngineKit *agoraRtcEngine;
@property(nonatomic, assign) BOOL isFrontCamera;
// Mirrors the Android controller's streamingRequested / inChannel / frameCount
// state (VideoRawDataController.kt:35-37).
@property(nonatomic, assign) BOOL streamingRequested;
@property(nonatomic, assign) BOOL inChannel;
@property(nonatomic, assign) NSInteger frameCount;

@end

@implementation VideoRawDataController

- (instancetype)initWith:(NSString *) appId {
    self = [super init];
    if (self) {
        AgoraRtcEngineConfig *config = [[AgoraRtcEngineConfig alloc] init];
        config.appId = appId;
        self.agoraRtcEngine = [AgoraRtcEngineKit sharedEngineWithConfig:config delegate:self];

        [self.agoraRtcEngine enableVideo];
        [self.agoraRtcEngine setLocalVideoMirrorMode:AgoraVideoMirrorModeDisabled];

        self.isFrontCamera = YES;
        self.streamingRequested = NO;
        self.inChannel = NO;
        self.frameCount = 0;

        [self configureEncoder];

        NSLog(@"VideoRawDataController initialized (Nosmai-owns-camera push model)");
    }

    return self;
}

// Previously this applied a default-constructed AgoraVideoEncoderConfiguration
// with only mirrorMode set, leaving dimensions/frameRate/bitrate/orientation at
// whatever a freshly allocated object yields — which can override Agora's own
// defaults rather than inherit them. Now it is an explicit 1:1 copy of Android.
- (void)configureEncoder {
    AgoraVideoEncoderConfiguration *encoderConfig =
        [[AgoraVideoEncoderConfiguration alloc]
             initWithSize:CGSizeMake(kStreamWidth, kStreamHeight)
                frameRate:AgoraVideoFrameRateFps30
                  bitrate:kStreamBitrateKbps
          orientationMode:AgoraVideoOutputOrientationModeFixedPortrait
               mirrorMode:AgoraVideoMirrorModeDisabled];
    encoderConfig.minBitrate = kStreamMinBitrateKbps;
    [self.agoraRtcEngine setVideoEncoderConfiguration:encoderConfig];
    NSLog(@"Agora encoder: %ldx%ld %ldkbps min=%ld",
          (long)kStreamWidth, (long)kStreamHeight,
          (long)kStreamBitrateKbps, (long)kStreamMinBitrateKbps);
}

- (intptr_t)getNativeHandle {
    return (intptr_t)[self.agoraRtcEngine getNativeHandle];
}

- (void)switchCamera {
    // Camera ownership belongs to the Nosmai preview in this model — Agora must
    // not drive the capture device. Matches Android's no-op switchCamera()
    // (VideoRawDataController.kt:190-192); we only track which lens is active.
    self.isFrontCamera = !self.isFrontCamera;
}

- (void)notifyCameraSwitch {
    self.isFrontCamera = !self.isFrontCamera;
}

#pragma mark - Streaming

- (BOOL)startStreaming:(NSString *)token
               channel:(NSString *)channelName
                   uid:(NSUInteger)uid {
    AgoraRtcEngineKit *engine = self.agoraRtcEngine;
    if (!engine) return NO;
    if (channelName.length == 0) return NO;

    // RE-ENTRANCY GUARD — same rationale as Android
    // (VideoRawDataController.kt:97-109): the Dart caller is driven by a socket
    // event the server can emit more than once. Without this, a second call
    // re-ran setup while the first join was still in flight, Agora rejected it
    // with -17, and the failure path then tore down the attempt that was about
    // to succeed — channel joined, producer off, black remote, no error.
    if (self.streamingRequested) {
        NSLog(@"startStreaming ignored: already streaming/joining channel=%@", channelName);
        return YES;
    }

    self.streamingRequested = YES;
    self.frameCount = 0;

    // Register as the video producer BEFORE joining, so the custom track exists
    // when the channel options ask to publish it.
    // useTexture:NO — we push CVPixelBuffers (AgoraVideoFrame format 12). iOS has
    // no equivalent of Android's EGL share-group texture path.
    [engine setExternalVideoSource:YES
                        useTexture:NO
                        sourceType:AgoraExternalVideoSourceTypeVideoFrame];
    [engine setClientRole:AgoraClientRoleBroadcaster];

    if (![self armNosmaiFrameCallback]) {
        // The camera plugin is absent or too old to expose the push API. Fail
        // loudly instead of joining a channel that would publish nothing.
        self.streamingRequested = NO;
        NSLog(@"❌ startStreaming: NosmaiFlutterPlugin does not expose "
              @"+setLiveStreamFrameCallback: — cannot publish filtered frames");
        return NO;
    }

    AgoraRtcChannelMediaOptions *options = [[AgoraRtcChannelMediaOptions alloc] init];
    options.clientRoleType = AgoraClientRoleBroadcaster;
    options.channelProfile = AgoraChannelProfileLiveBroadcasting;
    options.publishCameraTrack = NO;        // Nosmai owns the camera, not Agora
    options.publishCustomVideoTrack = YES;  // …and we push its filtered output
    options.publishCustomAudioTrack = YES;
    options.autoSubscribeVideo = YES;
    options.autoSubscribeAudio = YES;

    NSString *safeToken = (token.length > 0) ? token : nil;
    int result = [engine joinChannelByToken:safeToken
                                  channelId:channelName
                                        uid:uid
                               mediaOptions:options
                                joinSuccess:nil];
    if (result != 0) {
        self.streamingRequested = NO;
        DisarmNosmaiLiveFrames();
        NSLog(@"joinChannel failed result=%d", result);
        return NO;
    }
    return YES;
}

- (BOOL)stopStreaming {
    self.streamingRequested = NO;
    DisarmNosmaiLiveFrames();

    AgoraRtcEngineKit *engine = self.agoraRtcEngine;
    if (!engine) return YES;

    // Disarm the external source so a reused engine (the app keeps the engine
    // alive across streams) re-arms cleanly next time instead of inheriting a
    // stale registration. Mirrors Android stopStreaming (kt:164-175).
    [engine setExternalVideoSource:NO
                        useTexture:NO
                        sourceType:AgoraExternalVideoSourceTypeVideoFrame];

    int result = [engine leaveChannel:nil];
    self.inChannel = NO;
    return result == 0;
}

- (BOOL)armNosmaiFrameCallback {
    __weak typeof(self) weakSelf = self;
    return ArmNosmaiLiveFrames(^(CVPixelBufferRef pixelBuffer, double timestamp) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) return;
        [strongSelf pushFilteredBuffer:pixelBuffer timestamp:timestamp];
    });
}

- (void)pushFilteredBuffer:(CVPixelBufferRef)pixelBuffer timestamp:(double)timestamp {
    if (!pixelBuffer) return;

    // Gate on streamingRequested ONLY — deliberately not on inChannel.
    // inChannel is set exclusively from -rtcEngine:didJoinChannel: (:299 below).
    // Once the Flutter side builds its engine via
    // createAgoraRtcEngine(sharedNativeHandle:), the Dart/Iris layer attaches its
    // own delegate to this same native engine, and whether THIS ObjC delegate
    // still receives didJoinChannel is not knowable from code. If it does not,
    // inChannel stays NO forever and every frame is silently dropped — a joined
    // channel with a permanently black remote and no error anywhere.
    // streamingRequested is owned by this class's own start/stop path (set :143,
    // cleared :176/:194/:205), so it is immune to delegate delivery.
    if (!self.streamingRequested) return;

    AgoraRtcEngineKit *engine = self.agoraRtcEngine;
    if (!engine) return;

    // BUFFER LIFETIME: pushed synchronously on the SDK's frame-output thread. The
    // producer (nosmai_recording_sink.cc:596-599) does CVPixelBufferRetain →
    // callback(…) → CVPixelBufferRelease around this call, on a POOLED buffer
    // whose recycling is owned by RetireCompletedOutputs via pending_outputs_.
    // Whether -pushExternalVideoFrame consumes the buffer synchronously or
    // retains it for the encoder is UNVERIFIED. If it retains asynchronously, the
    // pool could recycle a buffer still queued in the encoder, giving intermittent
    // block corruption on the remote. We therefore retain across the push: this is
    // a refcount bump, not a copy, so it is cheap insurance on the hot path.
    CVPixelBufferRetain(pixelBuffer);

    size_t bufferWidth = CVPixelBufferGetWidth(pixelBuffer);
    size_t bufferHeight = CVPixelBufferGetHeight(pixelBuffer);

    AgoraVideoFrame *frame = [[AgoraVideoFrame alloc] init];
    frame.format = kAgoraVideoFormatCVPixelBuffer;
    frame.textureBuf = pixelBuffer;
    frame.time = CMTimeMakeWithSeconds(timestamp, 1000);

    // ROTATION — expected to evaluate to 0, because the source is already
    // PORTRAIT. The session preset names a 1280x720 FORMAT, but AVFoundation
    // rotates it: NosmaiSDK.mm:5708 sets
    // `param.videoOrientation = AVCaptureVideoOrientationPortrait`, applied to the
    // data-output connection at VideoCapturer.m:245. So buffers arrive upright at
    // 720x1280 and this ternary yields 0 into the FIXED_PORTRAIT 720x1280 encoder,
    // which is correct. (Do NOT "fix" this to a hardcoded 90 — that would rotate
    // already-correct portrait frames into landscape.)
    //
    // Front-camera mirroring is likewise already baked in by hardware
    // (VideoCapturer.m:240-243 sets videoMirrored = YES on the front connection),
    // which is why no bridge-side mirroring exists. That also makes the 90-vs-270
    // question moot: rotation is 0 on BOTH lenses.
    //
    // Kept as a dimension-derived expression rather than a literal 0 so it stays
    // correct if the SDK ever delivers landscape. The telemetry below prints the
    // actual size and rotation — verify it reads 720x1280 rotation=0 on device.
    frame.rotation = (bufferWidth > bufferHeight) ? 90 : 0;

    // videoTrackId 0 = the default custom video track, i.e. the one
    // ChannelMediaOptions.publishCustomVideoTrack publishes. The single-argument
    // -pushExternalVideoFrame: is deprecated in this Agora version.
    [engine pushExternalVideoFrame:frame videoTrackId:0];

    // Balances the retain above. If Agora retained the buffer for its encoder it
    // now holds its own reference; if it consumed it synchronously this is the
    // last release and the pool reclaims it as before.
    CVPixelBufferRelease(pixelBuffer);

    self.frameCount++;
    if (self.frameCount % 90 == 0) {
        NSLog(@"Pushed CVPixelBuffer frames=%ld size=%zux%zu rotation=%d",
              (long)self.frameCount, bufferWidth, bufferHeight, frame.rotation);
    }
}

- (void)dispose {
    @try {
        [self stopStreaming];
    } @catch (NSException *exception) {
        NSLog(@"dispose: stopStreaming threw: %@", exception.reason);
    }

    self.agoraRtcEngine = nil;
    [AgoraRtcEngineKit destroy];
}

#pragma mark - AgoraRtcEngineDelegate

- (void)rtcEngine:(AgoraRtcEngineKit *)engine
    didJoinChannel:(NSString *)channel
           withUid:(NSUInteger)uid
           elapsed:(NSInteger)elapsed {
    self.inChannel = YES;
    self.frameCount = 0;
    NSLog(@"Agora joined channel=%@ uid=%lu", channel, (unsigned long)uid);
}

- (void)rtcEngine:(AgoraRtcEngineKit *)engine didLeaveChannelWithStats:(AgoraChannelStats *)stats {
    self.inChannel = NO;
    // Only disarm when the app actually asked to stop. A leave callback can
    // arrive while a stream is still wanted (e.g. a rejected duplicate join
    // leaves a channel it never really held); disarming unconditionally there
    // would silence the producer for the stream that IS running.
    // Same guard as Android (VideoRawDataController.kt:54-61).
    if (!self.streamingRequested) {
        DisarmNosmaiLiveFrames();
    }
    NSLog(@"Agora left channel (streamingRequested=%d)", self.streamingRequested);
}

- (void)rtcEngine:(AgoraRtcEngineKit *)engine didOccurError:(AgoraErrorCode)errorCode {
    NSLog(@"Agora error=%ld", (long)errorCode);
}

@end

File: android/src/main/kotlin/com/example/nosmai_agora_bridge/VideoRawDataController.kt
package com.example.nosmai_agora_bridge

import android.content.Context
import android.util.Log
import com.nosmai.effect.api.NosmaiSDK
import io.agora.base.JavaI420Buffer
import io.agora.base.TextureBufferHelper
import io.agora.base.VideoFrame
import io.agora.base.internal.video.EglBase
import io.agora.rtc2.ChannelMediaOptions
import io.agora.rtc2.Constants
import io.agora.rtc2.IRtcEngineEventHandler
import io.agora.rtc2.RtcEngine
import io.agora.rtc2.RtcEngineConfig
import io.agora.rtc2.gl.EglBaseProvider
import io.agora.rtc2.video.VideoEncoderConfiguration
import java.nio.ByteBuffer
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class VideoRawDataController(context: Context, myAppId: String) {
    companion object {
        private const val TAG = "NosmaiAgoraBridge"
        private const val STREAM_WIDTH = 720
        private const val STREAM_HEIGHT = 1280
        private const val STREAM_BITRATE_KBPS = 2500
        private const val STREAM_MIN_BITRATE_KBPS = 1200
    }

    private var rtcEngine: RtcEngine? = null
    private var textureHelper: TextureBufferHelper? = null
    private var textureBridge: AgoraTextureBridge? = null
    private var pixelExecutor: ExecutorService? = null
    private var textureMode = false
    private var inChannel = false
    private var streamingRequested = false
    private var frameCount = 0

    private val rtcEventHandler = object : IRtcEngineEventHandler() {
        override fun onJoinChannelSuccess(channel: String?, uid: Int, elapsed: Int) {
            super.onJoinChannelSuccess(channel, uid, elapsed)
            inChannel = true
            frameCount = 0
            Log.i(TAG, "Agora joined channel=$channel uid=$uid textureMode=$textureMode")

            if (streamingRequested && textureMode) {
                armTextureCallback()
            }
        }

        override fun onLeaveChannel(stats: RtcStats?) {
            super.onLeaveChannel(stats)
            inChannel = false
            // Only disarm if the app actually asked to stop. A leave callback can
            // arrive while a stream is still wanted (e.g. a rejected duplicate
            // join leaves the channel it never really held); clearing callbacks
            // unconditionally there would silence the producer for the stream
            // that IS running.
            if (!streamingRequested) {
                clearStreamCallbacks()
            }
            Log.i(TAG, "Agora left channel (streamingRequested=$streamingRequested)")
        }

        override fun onError(err: Int) {
            super.onError(err)
            Log.e(TAG, "Agora error=$err")
        }
    }

    init {
        rtcEngine = RtcEngine.create(RtcEngineConfig().apply {
            mAppId = myAppId
            mContext = context.applicationContext
            mEventHandler = rtcEventHandler
        })
        pixelExecutor = Executors.newSingleThreadExecutor()

        rtcEngine?.enableVideo()
        rtcEngine?.setLocalVideoMirrorMode(Constants.VIDEO_MIRROR_MODE_DISABLED)
        // Register the Nosmai↔Agora EGL share context ONCE, up-front. This MUST
        // happen before the Nosmai GL context is created (else the share group is
        // wrong → black remote), and it is a one-time SDK registration — never
        // un-set per stream. Only the TextureBufferHelper (the GL thread that
        // contends the shared context) is created/destroyed per stream. See
        // registerShareContext() / createTextureHelper() / releaseTextureHelper().
        registerShareContext()
        configureEncoder()
    }

    fun nativeHandle(): Long = rtcEngine?.nativeHandle ?: 0L

    fun startStreaming(token: String?, channelName: String, uid: Int): Boolean {
        val engine = rtcEngine ?: return false
        if (channelName.isBlank()) return false

        // RE-ENTRANCY GUARD. The Dart caller is driven by a socket event
        // ("stream-started") which the server can emit more than once, so this
        // could be entered twice milliseconds apart. Without the guard the second
        // call re-ran the whole setup while the first joinChannel was still in
        // flight; Agora rejected it with -17 (ERR_JOIN_CHANNEL_REJECTED) and the
        // failure path below then cleared the texture callback and reset the
        // render mode to PREVIEW_ONLY — tearing down the FIRST attempt, which was
        // about to succeed. The channel joined, but the frame producer was off:
        // zero frames pushed, black remote, no error and no frozen frame.
        if (streamingRequested) {
            Log.w(TAG, "startStreaming ignored: already streaming/joining channel=$channelName")
            return true
        }

        streamingRequested = true
        frameCount = 0
        clearStreamCallbacks()
        // Create the TextureBufferHelper GL thread FRESH for this stream. Doing it
        // per-stream (instead of once at init) is the fix for the 2nd-go-live
        // freeze→black: previously the helper GL thread + its Nosmai EGL share-group
        // binding persisted across stopStreaming and kept contending the single
        // Nosmai GL worker, so on the next go-live the platform-thread surface-
        // release SyncRunWithContext head-of-line-blocked ~2.8s behind the saturated,
        // share-context-contended worker → frozen frame → permanent black.
        createTextureHelper()
        NosmaiSDK.setRenderMode(NosmaiSDK.RenderMode.DUAL_OUTPUT)

        engine.setExternalVideoSource(
            true,
            textureMode,
            Constants.ExternalVideoSourceType.VIDEO_FRAME
        )
        engine.setClientRole(Constants.CLIENT_ROLE_BROADCASTER)

        if (textureMode) {
            Log.i(TAG, "Streaming via zero-readback texture path; callback arms after join")
        } else {
            Log.w(TAG, "Texture mode unavailable; streaming via CPU I420 fallback")
            armPixelCallback()
        }

        val options = ChannelMediaOptions().apply {
            clientRoleType = Constants.CLIENT_ROLE_BROADCASTER
            channelProfile = Constants.CHANNEL_PROFILE_LIVE_BROADCASTING
            publishCameraTrack = false
            publishCustomVideoTrack = true
            publishCustomAudioTrack = true
            autoSubscribeVideo = true
            autoSubscribeAudio = true
        }

        val safeToken = token?.takeIf { it.isNotBlank() }
        val result = engine.joinChannel(safeToken, channelName, uid, options)
        if (result != 0) {
            streamingRequested = false
            clearStreamCallbacks()
            NosmaiSDK.setRenderMode(NosmaiSDK.RenderMode.PREVIEW_ONLY)
            Log.e(TAG, "joinChannel failed result=$result")
            return false
        }
        return true
    }

    fun stopStreaming(): Boolean {
        streamingRequested = false
        clearStreamCallbacks()
        NosmaiSDK.setRenderMode(NosmaiSDK.RenderMode.PREVIEW_ONLY)
        // Disarm the external video source so a reused engine (the app keeps the
        // engine alive across streams) re-arms cleanly on the next startStreaming
        // instead of inheriting a stale external-source registration.
        try {
            rtcEngine?.setExternalVideoSource(
                false,
                textureMode,
                Constants.ExternalVideoSourceType.VIDEO_FRAME
            )
        } catch (t: Throwable) {
            Log.w(TAG, "stopStreaming: disarm external source failed: ${t.message}")
        }
        val result = rtcEngine?.leaveChannel() ?: 0
        inChannel = false
        // Tear down the TextureBufferHelper GL thread + its Nosmai share-group
        // binding NOW (not at dispose). This is the core of the 2nd-go-live freeze
        // fix: the contending helper GL thread must NOT survive across the
        // go-live/end-live cycle. The Nosmai render mode is already back to
        // PREVIEW_ONLY (above), so after this the Nosmai GL worker runs uncontended
        // and the next createSinkPreviewTexture's platform-thread Sync drains fast.
        // The share-context REGISTRATION stays (one-time); only the helper thread
        // goes. A fresh helper is created in the next startStreaming.
        releaseTextureHelper()
        return result == 0
    }

    fun switchCamera() {
        // Camera ownership belongs to NosmaiCameraPreview in texture mode.
    }

    fun dispose() {
        try {
            stopStreaming()
        } catch (_: Throwable) {
        }

        // stopStreaming already tore down the helper (releaseTextureHelper); call
        // again defensively in case dispose is reached without a prior stream.
        releaseTextureHelper()
        textureMode = false

        pixelExecutor?.shutdown()
        pixelExecutor = null

        rtcEngine = null
        RtcEngine.destroy()
    }

    // Whether the one-time Nosmai↔Agora EGL share-context registration succeeded.
    // The share context is a one-time SDK registration (must precede the Nosmai GL
    // context); the TextureBufferHelper GL thread is created/destroyed per stream.
    private var shareContextReady = false

    /**
     * ONE-TIME: load libnosmai and register the Agora EGL context as the Nosmai
     * share context. Must run before the Nosmai GL context is created. Does NOT
     * create the TextureBufferHelper (that is per-stream — see createTextureHelper).
     */
    private fun registerShareContext() {
        try {
            try {
                System.loadLibrary("nosmai")
                Log.i(TAG, "libnosmai loaded before Agora share-context registration")
            } catch (t: Throwable) {
                shareContextReady = false
                Log.w(TAG, "Early libnosmai load failed; CPU fallback will be used", t)
                return
            }

            val agoraContext: EglBase.Context =
                EglBaseProvider.instance().getRootEglBase().getEglBaseContext()
            val nativeHandle = agoraContext.getNativeEglContext()
            NosmaiSDK.setAgoraShareContext(nativeHandle)
            shareContextReady = nativeHandle != 0L
            Log.i(TAG, "Share-context registered: shareCtx=$nativeHandle ready=$shareContextReady")
        } catch (t: Throwable) {
            shareContextReady = false
            Log.w(TAG, "Share-context registration failed; CPU fallback will be used", t)
        }
    }

    /**
     * PER-STREAM: create the TextureBufferHelper GL thread + the texture bridge.
     * Idempotent (reuses an existing helper if somehow still alive). Called from
     * startStreaming so the contending helper thread only exists while streaming.
     */
    private fun createTextureHelper() {
        val engine = rtcEngine ?: return
        if (!shareContextReady) {
            textureMode = false
            Log.w(TAG, "createTextureHelper: share context not ready; CPU fallback")
            return
        }
        if (textureHelper != null) {
            // Already have one (e.g. defensive double-call) — keep it.
            textureMode = textureBridge != null
            return
        }
        try {
            val agoraContext: EglBase.Context =
                EglBaseProvider.instance().getRootEglBase().getEglBaseContext()
            val helper = TextureBufferHelper.create("nosmai-flutter-agora", agoraContext)
            textureHelper = helper
            textureMode = helper != null
            if (textureMode) {
                textureBridge = AgoraTextureBridge(helper!!, engine)
            }
            Log.i(TAG, "createTextureHelper: helper=${textureHelper != null} textureMode=$textureMode")
        } catch (t: Throwable) {
            textureMode = false
            textureHelper = null
            textureBridge = null
            Log.w(TAG, "createTextureHelper failed; CPU fallback will be used", t)
        }
    }

    /**
     * PER-STREAM: tear down the TextureBufferHelper GL thread + bridge so it stops
     * contending the Nosmai GL worker via the shared EGL context. Safe to call when
     * already torn down. The share-context registration is NOT touched.
     */
    private fun releaseTextureHelper() {
        try { textureBridge?.release() } catch (_: Throwable) {}
        textureBridge = null
        try { textureHelper?.dispose() } catch (_: Throwable) {}
        textureHelper = null
        Log.i(TAG, "releaseTextureHelper: helper GL thread torn down (share-context kept)")
    }

    private fun configureEncoder() {
        val config = VideoEncoderConfiguration(
            VideoEncoderConfiguration.VideoDimensions(STREAM_WIDTH, STREAM_HEIGHT),
            VideoEncoderConfiguration.FRAME_RATE.FRAME_RATE_FPS_30,
            STREAM_BITRATE_KBPS,
            VideoEncoderConfiguration.ORIENTATION_MODE.ORIENTATION_MODE_FIXED_PORTRAIT
        )
        config.minBitrate = STREAM_MIN_BITRATE_KBPS
        config.mirrorMode = VideoEncoderConfiguration.MIRROR_MODE_TYPE.MIRROR_MODE_DISABLED
        rtcEngine?.setVideoEncoderConfiguration(config)
        Log.i(
            TAG,
            "Agora encoder: ${STREAM_WIDTH}x$STREAM_HEIGHT ${STREAM_BITRATE_KBPS}kbps min=$STREAM_MIN_BITRATE_KBPS"
        )
    }

    private fun armTextureCallback() {
        NosmaiSDK.setTextureFrameCallback { texId, width, height, timestampNs, fence ->
            val bridge = textureBridge
            if (!inChannel || bridge == null || !textureMode) {
                NosmaiSDK.releaseStreamSlot(texId)
                return@setTextureFrameCallback
            }
            bridge.pushCopy(texId, width, height, timestampNs) {
                NosmaiSDK.releaseStreamSlot(texId)
            }
            frameCount++
            if (frameCount % 90 == 0) {
                Log.i(TAG, "Pushed texture frames=$frameCount size=${width}x$height")
            }
        }
    }

    private fun armPixelCallback() {
        NosmaiSDK.setFrameCallback { frame ->
            if (!inChannel || frame?.pixelBuffer == null) return@setFrameCallback
            pushPixelFrame(
                frame.pixelBuffer,
                frame.width,
                frame.height,
                frame.timestampNs,
                frame.format,
                0
            )
        }
    }

    private fun clearStreamCallbacks() {
        try {
            NosmaiSDK.setTextureFrameCallback(null)
        } catch (_: Throwable) {
        }
        try {
            NosmaiSDK.setFrameCallback(null)
        } catch (_: Throwable) {
        }
    }

    private fun pushPixelFrame(
        frameData: ByteArray,
        width: Int,
        height: Int,
        timestampNs: Long,
        format: Int,
        rotation: Int
    ) {
        val engine = rtcEngine ?: return
        val executor = pixelExecutor ?: return
        if (!inChannel || format != 1) return

        executor.execute {
            try {
                val ySize = width * height
                val uvWidth = (width + 1) / 2
                val uvHeight = (height + 1) / 2
                val uSize = uvWidth * uvHeight
                val vSize = uSize
                val expectedSize = ySize + uSize + vSize
                if (frameData.size < expectedSize) return@execute

                val dataY = ByteBuffer.allocateDirect(ySize)
                val dataU = ByteBuffer.allocateDirect(uSize)
                val dataV = ByteBuffer.allocateDirect(vSize)
                dataY.put(frameData, 0, ySize)
                dataU.put(frameData, ySize, uSize)
                dataV.put(frameData, ySize + uSize, vSize)
                dataY.rewind()
                dataU.rewind()
                dataV.rewind()

                val buffer = JavaI420Buffer.wrap(
                    width,
                    height,
                    dataY,
                    width,
                    dataU,
                    uvWidth,
                    dataV,
                    uvWidth,
                    null
                )
                val videoFrame = VideoFrame(buffer, rotation, timestampNs)
                engine.pushExternalVideoFrame(videoFrame)
                videoFrame.release()
                frameCount++
                if (frameCount % 90 == 0) {
                    Log.i(TAG, "Pushed I420 frames=$frameCount size=${width}x$height")
                }
            } catch (t: Throwable) {
                Log.e(TAG, "pushPixelFrame failed", t)
            }
        }
    }
}
